File: app/src/main/java/com/gmail/fattazzo/formula1world/service/DataService.java
package com.gmail.fattazzo.formula1world.service;

import android.content.Context;
import android.graphics.Color;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;

import com.gmail.fattazzo.formula1world.R;
import com.gmail.fattazzo.formula1world.domain.F1Constructor;
import com.gmail.fattazzo.formula1world.domain.F1ConstructorStandings;
import com.gmail.fattazzo.formula1world.domain.F1Driver;
import com.gmail.fattazzo.formula1world.domain.F1DriverStandings;
import com.gmail.fattazzo.formula1world.domain.F1LapTime;
import com.gmail.fattazzo.formula1world.domain.F1PitStop;
import com.gmail.fattazzo.formula1world.domain.F1Qualification;
import com.gmail.fattazzo.formula1world.domain.F1Race;
import com.gmail.fattazzo.formula1world.domain.F1Result;
import com.gmail.fattazzo.formula1world.domain.F1Season;
import com.gmail.fattazzo.formula1world.ergast.DataCache;
import com.gmail.fattazzo.formula1world.ergast.Ergast;
import com.gmail.fattazzo.formula1world.ergast.IDataService;
import com.gmail.fattazzo.formula1world.ergast.imagedb.service.LocalDBDataService;
import com.gmail.fattazzo.formula1world.ergast.json.service.OnlineDataService;
import com.gmail.fattazzo.formula1world.settings.ApplicationPreferenceManager;
import com.gmail.fattazzo.formula1world.utils.Utils;

import org.androidannotations.annotations.Bean;
import org.androidannotations.annotations.EBean;
import org.androidannotations.annotations.RootContext;
import org.apache.commons.collections4.CollectionUtils;
import org.apache.commons.lang3.ObjectUtils;
import org.apache.commons.lang3.time.DateFormatUtils;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

/**
 * @author fattazzo
 *         <p/>
 *         date: 03/06/17
 */
@EBean(scope = EBean.Scope.Singleton)
public class DataService implements IDataService {

    @RootContext
    Context context;

    @Bean
    Utils utils;

    @Bean
    Ergast ergast;

    @Bean
    OnlineDataService onlineDataService;

    @Bean
    LocalDBDataService localDBDataService;

    @Bean
    DataCache dataCache;

    @Bean
    ApplicationPreferenceManager preferenceManager;

    private List<Integer> availableSeasons;

    // ------------------- Data Cache Actions -------------------
    public void clearCache() {
        dataCache.clearAll();
    }

    public void clearDriverStandingsCache() {
        dataCache.clearDriverStandings();
    }

    public void clearConstructorStandingsCache() {
        dataCache.clearConstructorStandings();
    }

    public void clearDriversCache() {
        dataCache.clearDrivers();
    }

    public void clearConstructorsCache() {
        dataCache.clearConstructors();
    }

    public void clearRacesCache() {
        dataCache.clearRaces();
    }

    public void clearRaceResultsCache(F1Race race) {
        dataCache.clearRaceResults(race);
    }

    public void clearRaceQualifications(F1Race race) {
        dataCache.clearRaceQualifications(race);
    }

    public void clearDriverRaceResultsCache(F1Driver driver) {
        dataCache.clearDriverRaceResults(driver);
    }

    public void clearConstructorRaceResultsCache(F1Constructor constructor) {
        dataCache.clearConstructorRaceResults(constructor);
    }

    public void clearRacePitStopsCache(F1Race race) {
        dataCache.clearRacePitStops(race);
    }

    public void clearColorsCache() {
        dataCache.clearConstructorColors();
        dataCache.clearDriverColors();
    }
    // ----------------------------------------------------------


    @Override
    public F1Season loadSeason(int year) {
        return getDataServiceImpl().loadSeason(year);
    }

    public List<Integer> getAvailableSeasons() {
        if (availableSeasons == null) {
            int currentYear = Calendar.getInstance().get(Calendar.YEAR);

            availableSeasons = new ArrayList<>();
            for (int i = currentYear; i >= 1950; i--) {
                availableSeasons.add(i);
            }
        }
        return availableSeasons;
    }

    @Override
    public void updateSeason(F1Season season) {
        getDataServiceImpl().updateSeason(season);
    }

    private IDataService getDataServiceImpl() {

        int season = getSelectedSeasons();
        boolean dbSeasonFound = localDBDataService.loadSeason(season) != null;

        // local db seasons win vs online seasons
        if (dbSeasonFound) {
            return localDBDataService;
        } else {
            return onlineDataService;
        }
    }

    public int getSelectedSeasons() {
        int season;
        try {
            if (ergast.getSeason() == Ergast.CURRENT_SEASON) {
                season = Calendar.getInstance().get(Calendar.YEAR);
            } else {
                season = ergast.getSeason();
            }
        } catch (Exception e) {
            season = -1;
        }
        return season;
    }

    @NonNull
    @Override
    public synchronized List<F1Driver> loadDrivers() {
        List<F1Driver> drivers = dataCache.getDrivers();
        if (CollectionUtils.isEmpty(drivers)) {
            drivers = localDBDataService.loadDrivers();
            dataCache.setDrivers(drivers);
        }
        return drivers;
    }

    @NonNull
    @Override
    public synchronized List<F1Constructor> loadConstructors() {
        List<F1Constructor> constructors = dataCache.getConstructors();
        if (CollectionUtils.isEmpty(constructors)) {
            constructors = localDBDataService.loadConstructors();
            dataCache.setConstructors(constructors);
        }
        return constructors;
    }

    @NonNull
    @Override
    public synchronized List<F1Result> loadDriverRacesResult(F1Driver driver) {
        List<F1Result> results = dataCache.getDriverRaceResults(driver);
        if (CollectionUtils.isEmpty(results)) {
            results = getDataServiceImpl().loadDriverRacesResult(driver);
            dataCache.setDriverRaceResults(driver, results);
        }
        return results;
    }

    @NonNull
    @Override
    public synchronized List<F1Result> loadConstructorRacesResult(F1Constructor constructor) {
        List<F1Result> results = dataCache.getConstructorRaceResults(constructor);
        if (CollectionUtils.isEmpty(results)) {
            results = getDataServiceImpl().loadConstructorRacesResult(constructor);
            dataCache.setConstructorRaceResults(constructor, results);
        }
        return results;
    }

    /**
     * Load current race scheduled.
     *
     * @return current race
     */
    @Nullable
    public synchronized F1Race loadCurrentSchedule() {
        List<F1Race> races = loadRaces();

        Calendar currentDateEnd = Calendar.getInstance();
        currentDateEnd.add(Calendar.HOUR_OF_DAY, 2);
        String currentDate = DateFormatUtils.format(currentDateEnd, "yyyy-MM-dd'T'HH:mm:ss");

        for (F1Race race : CollectionUtils.emptyIfNull(races)) {
            String scheudleDateUTC = race.date + "T" + race.time;
            String scheduleDateLocal = utils.convertUTCDateToLocal(scheudleDateUTC, "yyyy-MM-dd'T'HH:mm:ss", "yyyy-MM-dd'T'HH:mm:ss");

            if (scheduleDateLocal.compareTo(currentDate) >= 0) {
                return race;
            }
        }

        return null;
    }

    @NonNull
    @Override
    public synchronized List<F1DriverStandings> loadDriverStandings() {
        List<F1DriverStandings> driverStandings = dataCache.getDriverStandings();
        if (CollectionUtils.isEmpty(driverStandings)) {
            driverStandings = getDataServiceImpl().loadDriverStandings();
            dataCache.setDriverStandings(driverStandings);
        }
        return driverStandings;
    }

    @Nullable
    @Override
    public synchronized F1DriverStandings loadDriverLeader() {

        // First get from cache
        List<F1DriverStandings> standings = loadDriverStandings();
        for (F1DriverStandings standing : standings) {
            if (standing.position == 1) {
                return standing;
            }
        }
        return getDataServiceImpl().loadDriverLeader();
    }

    @NonNull
    @Override
    public synchronized List<F1ConstructorStandings> loadConstructorStandings() {
        List<F1ConstructorStandings> constructorStandings = dataCache.getConstructorStandings();
        if (CollectionUtils.isEmpty(constructorStandings)) {
            constructorStandings = getDataServiceImpl().loadConstructorStandings();
            dataCache.setConstructorStandings(constructorStandings);
        }
        return constructorStandings;
    }

    @NonNull
    @Override
    public synchronized List<F1Race> loadRaces() {
        List<F1Race> races = dataCache.getRaces();
        if (CollectionUtils.isEmpty(races)) {
            races = localDBDataService.loadRaces();
            dataCache.setRaces(races);
        }
        return races;
    }

    @NonNull
    @Override
    public synchronized List<F1Result> loadRaceResult(F1Race race) {
        List<F1Result> results = dataCache.getRaceResultsCache(race);
        if (CollectionUtils.isEmpty(results)) {
            results = localDBDataService.loadRaceResult(race);
            if (CollectionUtils.isEmpty(results)) {
                results = onlineDataService.loadRaceResult(race);
            }
            dataCache.setRaceResults(race, results);
        }
        return results;
    }

    @NonNull
    @Override
    public synchronized List<F1Qualification> loadQualification(F1Race race) {
        List<F1Qualification> qualifications = dataCache.getRaceQualificationsCache(race);
        if (CollectionUtils.isEmpty(qualifications)) {
            qualifications = localDBDataService.loadQualification(race);
            if (CollectionUtils.isEmpty(qualifications)) {
                qualifications = onlineDataService.loadQualification(race);
            }
            dataCache.setRaceQualifications(race, qualifications);
        }
        return qualifications;
    }

    @NonNull
    @Override
    public List<F1PitStop> loadPitStops(F1Race race) {
        List<F1PitStop> pitStops = dataCache.getRacePitStops(race);
        if (CollectionUtils.isEmpty(pitStops)) {
            pitStops = localDBDataService.loadPitStops(race);
            if (CollectionUtils.isEmpty(pitStops)) {
                pitStops = onlineDataService.loadPitStops(race);
            }
            dataCache.setRacePitStops(race, pitStops);
        }
        return pitStops;
    }

    @Override
    public List<F1LapTime> loadLaps(@NonNull F1Race race, @NonNull F1Driver driver) {
        List<F1LapTime> lapTimes = dataCache.getRaceLapTimes(race, driver);
        if (CollectionUtils.isEmpty(lapTimes)) {
            if (!hasLocalLapsData(race)) {
                lapTimes = onlineDataService.loadLaps(race, driver);
            } else {
                lapTimes = localDBDataService.loadLaps(race, driver);
            }
            if (CollectionUtils.isNotEmpty(lapTimes)) {
                dataCache.setRaceLapTimes(race, driver, lapTimes);
            }
        }
        return lapTimes;
    }

    public int loadContructorColor(F1Constructor constructor) {
        Integer color = dataCache.getConstructorColor(constructor);
        if (constructor != null && color == null) {
            // try from db
            color = localDBDataService.loadContructorColor(constructor);
            if (color == null) {
                // try from standard
                try {
                    int colorId = context.getResources().getIdentifier(constructor.constructorRef, "color", context.getPackageName());
                    color = Color.parseColor("#" + Integer.toHexString(context.getResources().getColor(colorId)));
                } catch (Exception e) {
                    color = null;
                }
            }

            if (color != null) {
                dataCache.setConstructorColor(constructor, color);
            }
        }

        return ObjectUtils.defaultIfNull(color, android.R.color.transparent);
    }

    public int loadDriverColor(F1Driver driver) {
        Integer color = dataCache.getDriverColor(driver);
        if (driver != null && color == null) {
            // try from db
            color = localDBDataService.loadDriverColor(driver);
            if (color == null) {
                // try from standard
                try {
                    F1Constructor constructor = localDBDataService.loadConstructor(driver);
                    int colorId = context.getResources().getIdentifier(constructor.constructorRef, "color", context.getPackageName());
                    color = Color.parseColor("#" + Integer.toHexString(context.getResources().getColor(colorId)));
                } catch (Exception e) {
                    color = null;
                }
            }

            if (color != null) {
                dataCache.setDriverColor(driver, color);
            }
        }

        return ObjectUtils.defaultIfNull(color, ContextCompat.getColor(context,R.color.background_color_dark));
    }

    public boolean hasLocalLapsData(F1Race race) {
        return localDBDataService.hasLocalLapsData(race);
    }
}

File: app/src/main/kotlin/com/gmail/fattazzo/formula1world/service/DataService.kt
/*
 * Project: total-gp-world
 * File: DataService.kt
 *
 * Created by fattazzo
 * Copyright © 2018 Gianluca Fattarsi. All rights reserved.
 *
 * MIT License
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.gmail.fattazzo.formula1world.service

import android.content.Context
import android.graphics.Color
import android.support.v4.content.ContextCompat
import com.gmail.fattazzo.formula1world.R
import com.gmail.fattazzo.formula1world.domain.*
import com.gmail.fattazzo.formula1world.ergast.DataCache
import com.gmail.fattazzo.formula1world.ergast.Ergast
import com.gmail.fattazzo.formula1world.ergast.IDataService
import com.gmail.fattazzo.formula1world.ergast.imagedb.service.LocalDBDataService
import com.gmail.fattazzo.formula1world.ergast.json.service.OnlineDataService
import com.gmail.fattazzo.formula1world.settings.ApplicationPreferenceManager
import com.gmail.fattazzo.formula1world.utils.Utils
import org.androidannotations.annotations.Bean
import org.androidannotations.annotations.EBean
import org.androidannotations.annotations.RootContext
import org.apache.commons.collections4.CollectionUtils
import org.apache.commons.lang3.ObjectUtils
import org.apache.commons.lang3.time.DateFormatUtils
import java.util.*

/**
 * @author fattazzo
 *
 *
 * date: 03/06/17
 */
@EBean(scope = EBean.Scope.Singleton)
open class DataService : IDataService {

    @RootContext
    lateinit internal var context: Context

    @Bean
    lateinit internal var utils: Utils

    @Bean
    lateinit internal var ergast: Ergast

    @Bean
    lateinit internal var onlineDataService: OnlineDataService

    @Bean
    lateinit internal var localDBDataService: LocalDBDataService

    @Bean
    lateinit internal var dataCache: DataCache

    @Bean
    lateinit internal var preferenceManager: ApplicationPreferenceManager

    private var availableSeasons: MutableList<Int>? = null

    private// local db seasons win vs online seasons
    val dataServiceImpl: IDataService
        get() {

            val season = selectedSeasons
            val dbSeasonFound = localDBDataService.loadSeason(season) != null
            return if (dbSeasonFound) {
                localDBDataService
            } else {
                onlineDataService
            }
        }

    val selectedSeasons: Int
        get() {
            return try {
                if (ergast.season == Ergast.CURRENT_SEASON) {
                    Calendar.getInstance().get(Calendar.YEAR)
                } else {
                    ergast.season
                }
            } catch (e: Exception) {
                -1
            }
        }

    // ------------------- Data Cache Actions -------------------
    fun clearCache() {
        dataCache.clearAll()
    }

    fun clearDriverStandingsCache() {
        dataCache.clearDriverStandings()
    }

    fun clearConstructorStandingsCache() {
        dataCache.clearConstructorStandings()
    }

    fun clearDriversCache() {
        dataCache.clearDrivers()
    }

    fun clearConstructorsCache() {
        dataCache.clearConstructors()
    }

    fun clearRacesCache() {
        dataCache.clearRaces()
    }

    fun clearRaceResultsCache(race: F1Race) {
        dataCache.clearRaceResults(race)
    }

    fun clearRaceQualifications(race: F1Race) {
        dataCache.clearRaceQualifications(race)
    }

    fun clearDriverRaceResultsCache(driver: F1Driver) {
        dataCache.clearDriverRaceResults(driver)
    }

    fun clearConstructorRaceResultsCache(constructor: F1Constructor) {
        dataCache.clearConstructorRaceResults(constructor)
    }

    fun clearRacePitStopsCache(race: F1Race) {
        dataCache.clearRacePitStops(race)
    }

    fun clearColorsCache() {
        dataCache.clearConstructorColors()
        dataCache.clearDriverColors()
    }
    // ----------------------------------------------------------


    override fun loadSeason(year: Int): F1Season? = dataServiceImpl.loadSeason(year)

    fun getAvailableSeasons(): List<Int> {
        if (availableSeasons == null) {
            var lastYear = localDBDataService.lastSeason()
            if (lastYear == null) {
                lastYear = Calendar.getInstance().get(Calendar.YEAR)
            }
            lastYear+=1

            availableSeasons = ArrayList()
            for (i in lastYear downTo 1950) {
                availableSeasons!!.add(i)
            }
        }
        return availableSeasons!!
    }

    override fun updateSeason(season: F1Season) {
        dataServiceImpl!!.updateSeason(season)
    }

    @Synchronized override fun loadDrivers(): List<F1Driver> {
        var drivers = dataCache.drivers
        if (CollectionUtils.isEmpty(drivers)) {
            drivers = localDBDataService.loadDrivers()
            if (CollectionUtils.isEmpty(drivers)) {
                drivers = onlineDataService.loadDrivers()
            }
            dataCache.drivers = drivers
        }
        return drivers
    }

    @Synchronized override fun loadConstructors(): List<F1Constructor> {
        var constructors = dataCache.constructors
        if (CollectionUtils.isEmpty(constructors)) {
            constructors = localDBDataService.loadConstructors()
            if (CollectionUtils.isEmpty(constructors)) {
                constructors = onlineDataService.loadConstructors()
            }
            dataCache.constructors = constructors
        }
        return constructors
    }

    @Synchronized override fun loadDriverRacesResult(driver: F1Driver): List<F1Result> {
        var results = dataCache.getDriverRaceResults(driver)
        if (CollectionUtils.isEmpty(results)) {
            results = dataServiceImpl!!.loadDriverRacesResult(driver)
            dataCache.setDriverRaceResults(driver, results)
        }
        return results
    }

    @Synchronized override fun loadConstructorRacesResult(constructor: F1Constructor): List<F1Result> {
        var results = dataCache.getConstructorRaceResults(constructor)
        if (CollectionUtils.isEmpty(results)) {
            results = dataServiceImpl!!.loadConstructorRacesResult(constructor)
            dataCache.setConstructorRaceResults(constructor, results)
        }
        return results
    }

    /**
     * Load current race scheduled.
     *
     * @return current race
     */
    @Synchronized
    fun loadCurrentSchedule(): F1Race? {
        val races = loadRaces()

        val currentDateEnd = Calendar.getInstance()
        currentDateEnd.add(Calendar.HOUR_OF_DAY, 2)
        val currentDate = DateFormatUtils.format(currentDateEnd, "yyyy-MM-dd'T'HH:mm:ss")

        for (race in CollectionUtils.emptyIfNull(races)) {
            val scheudleDateUTC = race.date + "T" + race.time
            val scheduleDateLocal = utils.convertUTCDateToLocal(scheudleDateUTC, "yyyy-MM-dd'T'HH:mm:ss", "yyyy-MM-dd'T'HH:mm:ss")

            if (scheduleDateLocal >= currentDate) {
                return race
            }
        }

        return null
    }

    @Synchronized override fun loadDriverStandings(): List<F1DriverStandings> {
        var driverStandings = dataCache.driverStandings
        if (CollectionUtils.isEmpty(driverStandings)) {
            driverStandings = dataServiceImpl!!.loadDriverStandings()
            dataCache.driverStandings = driverStandings
        }
        return driverStandings
    }

    @Synchronized override fun loadDriverLeader(): F1DriverStandings? {

        // First get from cache
        val standings = loadDriverStandings()
        for (standing in standings) {
            if (standing.position == 1) {
                return standing
            }
        }
        return dataServiceImpl!!.loadDriverLeader()
    }

    @Synchronized override fun loadConstructorStandings(): List<F1ConstructorStandings> {
        var constructorStandings = dataCache.constructorStandings
        if (CollectionUtils.isEmpty(constructorStandings)) {
            constructorStandings = dataServiceImpl!!.loadConstructorStandings()
            dataCache.constructorStandings = constructorStandings
        }
        return constructorStandings
    }

    @Synchronized override fun loadRaces(): List<F1Race> {
        var races = dataCache.races
        if (CollectionUtils.isEmpty(races)) {
            races = localDBDataService.loadRaces()
            if (CollectionUtils.isEmpty(races)) {
                races = onlineDataService.loadRaces()
            }
            dataCache.races = races
        }
        return races
    }

    @Synchronized override fun loadRaceResult(race: F1Race): List<F1Result> {
        var results = dataCache.getRaceResultsCache(race)
        if (CollectionUtils.isEmpty(results)) {
            results = localDBDataService.loadRaceResult(race)
            if (CollectionUtils.isEmpty(results)) {
                results = onlineDataService.loadRaceResult(race)
            }
            dataCache.setRaceResults(race, results)
        }
        return results
    }

    @Synchronized override fun loadQualification(race: F1Race): List<F1Qualification> {
        var qualifications = dataCache.getRaceQualificationsCache(race)
        if (CollectionUtils.isEmpty(qualifications)) {
            qualifications = localDBDataService.loadQualification(race)
            if (CollectionUtils.isEmpty(qualifications)) {
                qualifications = onlineDataService.loadQualification(race)
            }
            dataCache.setRaceQualifications(race, qualifications)
        }
        return qualifications
    }

    override fun loadPitStops(race: F1Race): List<F1PitStop> {
        var pitStops = dataCache.getRacePitStops(race)
        if (CollectionUtils.isEmpty(pitStops)) {
            pitStops = localDBDataService.loadPitStops(race)
            if (CollectionUtils.isEmpty(pitStops)) {
                pitStops = onlineDataService.loadPitStops(race)
            }
            dataCache.setRacePitStops(race, pitStops)
        }
        return pitStops
    }

    override fun loadLaps(race: F1Race, driver: F1Driver): List<F1LapTime> {
        var lapTimes = dataCache.getRaceLapTimes(race, driver)
        if (CollectionUtils.isEmpty(lapTimes)) {
            lapTimes = if (!hasLocalLapsData(race)) {
                onlineDataService.loadLaps(race, driver)
            } else {
                localDBDataService.loadLaps(race, driver)
            }
            if (CollectionUtils.isNotEmpty(lapTimes)) {
                dataCache.setRaceLapTimes(race, driver, lapTimes)
            }
        }
        return lapTimes
    }

    fun loadContructorColor(constructor: F1Constructor?): Int {
        var color = dataCache.getConstructorColor(constructor!!)
        if (constructor != null && color == null) {
            // try from db
            color = localDBDataService.loadContructorColor(constructor)
            if (color == null) {
                // try from standard
                color = try {
                    val colorId = context.resources.getIdentifier(constructor.constructorRef, "color", context.packageName)
                    Color.parseColor("#" + Integer.toHexString(context.resources.getColor(colorId)))
                } catch (e: Exception) {
                    null
                }

            }

            if (color != null) {
                dataCache.setConstructorColor(constructor, color)
            }
        }

        return ObjectUtils.defaultIfNull<Int>(color, android.R.color.transparent)
    }

    fun loadDriverColor(driver: F1Driver?): Int {
        var color = dataCache.getDriverColor(driver!!)
        if (driver != null && color == null) {
            // try from db
            color = localDBDataService.loadDriverColor(driver)
            if (color == null) {
                // try from standard
                color = try {
                    val constructor = localDBDataService.loadConstructor(driver)
                    val colorId = context.resources.getIdentifier(constructor!!.constructorRef, "color", context.packageName)
                    Color.parseColor("#" + Integer.toHexString(context.resources.getColor(colorId)))
                } catch (e: Exception) {
                    null
                }

            }

            if (color != null) {
                dataCache.setDriverColor(driver, color)
            }
        }

        return ObjectUtils.defaultIfNull<Int>(color, ContextCompat.getColor(context, R.color.background_color_dark))
    }

    fun hasLocalLapsData(race: F1Race): Boolean = localDBDataService.hasLocalLapsData(race)
}
